sql/builtin_functions.cc

```diff
@@ -2513,6 +2517,127 @@ BoundFunction upper(const std::vector<BoundSqlExpression> & args)
 
+
+BoundFunction levenshtein_distance(const std::vector<BoundSqlExpression> & args)
+{
+    if (args.size() != 2)
+        throw HttpReturnException(400, "levenshtein_distance function takes 2 arguments");
+
+     return {[=] (const std::vector<ExpressionValue> & args,
+                  const SqlRowScope & scope) -> ExpressionValue
+             {
+                ExcAssertEqual(args.size(), 2);
+                const auto query = args[0].getAtom().toUtf8String().rawString();
+                const auto target = args[1].getAtom().toUtf8String().rawString();
+
+                // start by testing easy edge cases
+                int bestScore = -1;
+                if(query.size() == 0 && target.size() == 0)
+                    bestScore = 0;
+                else if(query.size() == 0 || target.size() == 0)
+                    bestScore = max(query.size(), target.size());
+
+                if(bestScore != -1)
+                    return ExpressionValue(bestScore,
+                                           args[0].getEffectiveTimestamp());
+
+
+                bool singleThreaded = query.size() + target.size() < 1000;
+
+                unsigned char convQuery[query.size()];
+                unsigned char convTarget[target.size()];
+
+                mutex myMutex;
+                atomic<int> idx(0);
+                map<char, int> charIdx;
+
+                auto fct = [&] (const string & in, unsigned char * out) {
+                    auto strLen = in.size();
+                    for (int i = 0; i < strLen; ++i) {
+                        auto & currChar = in[i];
+                        auto it = charIdx.find(currChar);
+                        if (it == charIdx.end()) {
+
+                            auto doEmplace = [&] ()
+                                {
+                                    auto res = charIdx.emplace(currChar, idx ++);
+                                    ExcAssert(std::get<1>(res));
+                                    it = std::get<0>(res);
+                                };
+
+                            if(singleThreaded) {
+                                doEmplace();
+                            }
+                            else {
+                                lock_guard<mutex> lock(myMutex);
+                                it = charIdx.find(currChar);
+                                if (it == charIdx.end()) {
+                                    doEmplace();
+                                }
+                            }
+                        }
+                        out[i] = it->second;
+                    }
+                };
+
+                if(singleThreaded) {
+                    fct(query, &convQuery[0]);
+                    fct(target, &convTarget[0]);
+                }
+                else {
+                    thread queryThread(fct, query, &convQuery[0]);
+                    thread targetThread(fct, target, &convTarget[0]);
+                    queryThread.join();
```

Comment:
Oui en effet. Dans les fait de multithreader cette fonction est overkill et super lent puisque la majorité des string processés sont vraiment petites. Ce PR était pour montrer du code, pas pour être mergé as is. ;)